File: 05_01_limpieza_consumo.py
import pandas as pd
import os

# Crear carpeta output si no existe
if not os.path.exists('output'):
    os.makedirs('output')

def leer_banxico(archivo, nombre_col):
    print(f"Leyendo: {archivo}")
    # Detectar dónde empieza la data (buscando la palabra 'Fecha')
    try:
        with open(archivo, 'r', encoding='latin1') as f:
            lines = f.readlines()
            skip = 0
            for i, line in enumerate(lines):
                if 'Fecha' in line:
                    skip = i
                    break
        
        # Leer el dataframe
        df = pd.read_csv(archivo, encoding='latin1', skiprows=skip)
        # Tomar solo las primeras dos columnas (Fecha y Dato)
        df = df.iloc[:, [0, 1]]
        df.columns = ['fecha', nombre_col]
        
        # Limpiar fecha y convertir a número
        df['fecha'] = pd.to_datetime(df['fecha'], dayfirst=True, errors='coerce')
        df[nombre_col] = pd.to_numeric(df[nombre_col], errors='coerce')
        
        return df.dropna()
    except Exception as e:
        print(f"Error al leer {archivo}: {e}")
        return pd.DataFrame()

try:
    # 1. Procesar BANXICO
    df_fix = leer_banxico('input/tc_fix.csv', 'tc_fix')
    df_tasa = leer_banxico('input/tasa_nominal.csv', 'tasa_nom')
    df_inpc = leer_banxico('input/inflacion.csv', 'inpc')

    if df_fix.empty or df_tasa.empty or df_inpc.empty:
        raise ValueError("Uno de los archivos de Banxico no se pudo cargar.")

    # Unir datos de Banxico por mes
    df_inpc['fecha_m'] = df_inpc['fecha'].dt.to_period('M').dt.to_timestamp()
    df_tasa['fecha_m'] = df_tasa['fecha'].dt.to_period('M').dt.to_timestamp()
    df_fix['fecha_m'] = df_fix['fecha'].dt.to_period('M').dt.to_timestamp()

    # Calcular inflación mensual y tasa real
    df_inpc = df_inpc.sort_values('fecha_m')
    df_inpc['inf_mensual'] = df_inpc['inpc'].pct_change() * 100
    
    banxico = df_inpc.merge(df_tasa, on='fecha_m').merge(df_fix, on='fecha_m')
    
    # Tasa Real = Tasa Nominal - Inflación Anualizada
    banxico['inf_anual'] = ((1 + banxico['inf_mensual']/100)**12 - 1) * 100
    banxico['tasa_real'] = banxico['tasa_nom'] - banxico['inf_anual']
    
    # Agrupar por trimestre
    banxico['year'] = banxico['fecha_m'].dt.year
    banxico['trimestre'] = banxico['fecha_m'].dt.quarter
    banxico_trim = banxico.groupby(['year', 'trimestre'])[['tc_fix', 'tasa_real']].mean().reset_index()

    # 2. Procesar INEGI
    print("Leyendo: input/inegi_consumo.csv")
    # Intentar leer el inicio de la data en INEGI
    with open('input/inegi_consumo.csv', 'r', encoding='utf-16') as f:
        lines = f.readlines()
        skip_inegi = 0
        for i, line in enumerate(lines):
            if 'Periodo' in line:
                skip_inegi = i
                break

    df_inegi = pd.read_csv('input/inegi_consumo.csv', encoding='utf-16', skiprows=skip_inegi)
    # Limpiar nombres de columnas
    df_inegi = df_inegi.iloc[:, 0:4]
    df_inegi.columns = ['periodo', 'c_total', 'c_nacional', 'c_importado']
    
    # Limpiar el campo periodo (ej: "2000/01")
    df_inegi = df_inegi.dropna(subset=['c_total'])
    df_inegi['year'] = df_inegi['periodo'].str[:4].astype(int)
    df_inegi['trimestre'] = df_inegi['periodo'].str[5:7].astype(int)

    # 3. Unión Final
    df_final = df_inegi.merge(banxico_trim, on=['year', 'trimestre'])
    
    # Ordenar y guardar
    df_final = df_final[['year', 'trimestre', 'periodo', 'c_total', 'c_nacional', 'c_importado', 'tc_fix', 'tasa_real']]
    df_final.to_csv('output/consumo_limpio.csv', index=False)
    
    print("\n¡ÉXITO TOTAL!")
    print(f"Se ha generado 'output/consumo_limpio.csv' con {len(df_final)} filas.")
    print(df_final.head())

except Exception as e:
    print(f"\nERROR: {str(e)}")

File: 05_01_limpieza_consumo.R
# Script 05_01_limpieza_consumo.R (Versión 4 - Final)
# Objetivo: Limpiar y unir series con manejo avanzado de fechas y codificación.
# Autor: Gemini CLI

# Cargar librerías
if (!require("readr")) install.packages("readr")
if (!require("dplyr")) install.packages("dplyr")
if (!require("lubridate")) install.packages("lubridate")
library(readr)
library(dplyr)
library(lubridate)
# 1. Función para leer Banxico ignorando errores de formato y notas
leer_banxico <- function(archivo, nombre_col) {
  print(paste("Leyendo:", archivo))
  # Leer todo el archivo como texto primero
  lineas <- read_lines(archivo, locale = locale(encoding = "ISO-8859-1"))
  # Buscar la línea donde empiezan los datos (donde está el encabezado "Fecha")
  inicio <- which(lineas == '"Fecha","SF43718"' | lineas == '"Fecha","SF43936"' | lineas == '"Fecha","SP30562"' | grepl("Fecha", lineas))[1]
  # Leer desde ahí
  datos <- read_csv(archivo, skip = inicio - 1, show_col_types = FALSE, locale = locale(encoding = "ISO-8859-1"))
  colnames(datos) <- c("fecha_raw", "valor_raw")
  # Limpiar datos: Convertir fecha y valor, tirar lo que no sirva
  datos_limpios <- datos %>%
    mutate(
      # Extraer solo números y diagonales de la fecha
      fecha_limpia = gsub("[^0-9/]", "", as.character(fecha_raw)),
      fecha = dmy(fecha_limpia),
      # Convertir valor a número (maneja N/E o espacios)
      valor = as.numeric(as.character(valor_raw))
    ) %>%
    filter(!is.na(fecha), !is.na(valor)) %>%
    select(fecha, valor)
  colnames(datos_limpios) <- c("fecha", nombre_col)
  return(datos_limpios)
}
# 2. Ejecutar Limpieza
try({
  # Leer archivos de Banxico
  df_fix <- leer_banxico("input/tc_fix.csv", "tc_fix")
  df_tasa <- leer_banxico("input/tasa_nominal.csv", "tasa_nom")
  df_inpc <- leer_banxico("input/inflacion.csv", "inpc")
  # Calcular Tasa Real
  banxico <- df_inpc %>%
    mutate(fecha = floor_date(fecha, "month")) %>%
    arrange(fecha) %>%
    mutate(inf_mensual = (inpc / lag(inpc) - 1) * 100) %>%
    inner_join(mutate(df_tasa, fecha = floor_date(fecha, "month")), by = "fecha") %>%
    inner_join(mutate(df_fix, fecha = floor_date(fecha, "month")), by = "fecha") %>%
    mutate(
      inf_anual = ((1 + inf_mensual/100)^12 - 1) * 100,
      tasa_real = tasa_nom - inf_anual
    ) %>%
    mutate(year = year(fecha), trimestre = quarter(fecha)) %>%
    group_by(year, trimestre) %>%
    summarise(tc_fix = mean(tc_fix), tasa_real = mean(tasa_real), .groups = "drop")
  # Leer INEGI (Codificación UTF-16LE)
  print("Leyendo: input/inegi_consumo.csv")
  lineas_inegi <- read_lines("input/inegi_consumo.csv", locale = locale(encoding = "UTF-16LE"))
  inicio_inegi <- which(grepl("Periodo", lineas_inegi))[1]
  df_inegi <- read_csv("input/inegi_consumo.csv", skip = inicio_inegi - 1, 
                       locale = locale(encoding = "UTF-16LE"), show_col_types = FALSE)
  colnames(df_inegi)[1:4] <- c("periodo", "c_total", "c_nacional", "c_importado")
  df_inegi <- df_inegi %>%
    filter(!is.na(c_total)) %>%
    mutate(year = as.numeric(substr(periodo, 1, 4)), trimestre = as.numeric(substr(periodo, 6, 7)))
  # Unión Final
  df_final <- df_inegi %>%
    inner_join(banxico, by = c("year", "trimestre")) %>%
    select(year, trimestre, periodo, c_total, c_nacional, c_importado, tc_fix, tasa_real)
  if(!dir.exists("output")) dir.create("output")
  write_csv(df_final, "output/consumo_limpio.csv")
  print("¡LOGRADO! Archivo 'output/consumo_limpio.csv' generado con éxito.")
  print(head(df_final))
})